compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java

```diff
@@ -0,0 +1,1138 @@
+        }
+
+        return getExpectedType(bLangNode);
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(BasicLiteralNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        return getTypeFromBType(bLangNode.getBType());
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(AssignmentStatementNode node) {
+        return this.visit(node.varRef());
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(FieldAccessExpressionNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        return getExpectedType(bLangNode);
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(LetVariableDeclarationNode node) {
+        return this.visit(node.typedBindingPattern().bindingPattern());
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(ListenerDeclarationNode node) {
+        Optional<TypeDescriptorNode> typeDesc = node.typeDescriptor();
+        if (typeDesc.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return typeDesc.get().apply(this);
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(ModuleVariableDeclarationNode node) {
+        return this.visit(node.typedBindingPattern().bindingPattern());
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(VariableDeclarationNode node) {
+        if (node.typedBindingPattern().typeDescriptor().kind() == SyntaxKind.VAR_TYPE_DESC) {
+            Types types = semanticModel.types();
+            TypeBuilder builder = types.builder();
+            UnionTypeSymbol unionTypeSymbol = builder.UNION_TYPE.withMemberTypes(types.ANY, types.ERROR).build();
+            return Optional.of(unionTypeSymbol);
+        }
+
+        return this.visit(node.typedBindingPattern().bindingPattern());
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(BinaryExpressionNode node) {
+        if (!node.operator().isMissing() && node.operator().text().equals("*") || node.operator().text().equals("/")) {
+            return Optional.empty();
+        }
+
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (!(bLangNode instanceof BLangBinaryExpr)) {
+            return Optional.empty();
+        }
+
+        BLangBinaryExpr bLangBinaryExpr = (BLangBinaryExpr) bLangNode;
+        if (bLangBinaryExpr.rhsExpr.expectedType.getKind() == TypeKind.OTHER &&
+                bLangBinaryExpr.lhsExpr.expectedType.getKind() == TypeKind.OTHER) {
+            return Optional.empty();
+        } else if (bLangBinaryExpr.rhsExpr.expectedType.getKind() != TypeKind.OTHER) {
```

Comment:
Since we have a return statement in each block we can simply use only the `if` condition without the `else if` . (early returns)

Reply:
+1

Reply:
Also we can initialize `lhs` and `rhs` and use them

```java
BLangExpression rhs = bLangBinaryExpr.rhsExpr;
BLangExpression lhs = bLangBinaryExpr.lhsExpr;
```